### PR TITLE
Transformation performance optimizations

### PR DESCRIPTION
**Pull request by Scrub1492**
This change adds early bounds to the language transformation process, greatly improving language transformer performance. This might help reduce lookup lag and also cases like https://github.com/themoeway/yomitan/pull/584#issuecomment-1926538979.

### CODE REVIEW

**Comment by StefanVukovic99** [comment]
Well right now it only contains endsWith, so maybe instead of nesting 
```json
"matchConditions": {
     "endsWith": ["ば"]
}
```
just do it like:
```json
"sharedSuffix": ["ば"]
```
Or if the point of the nesting was other kinds of conditions, maybe use regex:
```json
"preconditions": ["ば$"]
```

**Comment by Scrub1492** [comment]
I don't know. Is the penalty for recompiling the regex every time worth it, I think a simple object is fine

**Comment by Scrub1492** [comment]
And more susceptible to Regex injections

**Comment by toasted-nutbread** [comment]
A thought: why even define these in the JSON file? A lot of this can be computed programmatically during construction. There is perhaps a special case for the situations where the suffix has a length=1 and has multiple characters, but even this can be detected. You could also entirely omit the suffix check if `[unique-one-character-suffixes].length === rules.length`.

**Comment by Scrub1492** [comment]
Do you mean creating a list of suffixes in the addDescriptor method? Can you provide an example, I couldn't quite imagine it.

**Comment by toasted-nutbread** [comment]
Yes, give me a bit and I can give a sample.

**Comment by Scrub1492** [comment]
The reason I defined those in the JSON was just so that each transformation case has more control over how they are matched. Other languages may have a different transforms.json where the suffixes are all different, then we can just mark endsWith as optional and add in another matcher for that language.

**Comment by toasted-nutbread** [comment]
If the suffixes are all different, there's not much utility in the pre-process step for that situation.

Here is some code that can compute the suffixes:
```js

/**
 * @param {string} text1
 * @param {string} text2
 * @returns {number}
 */
function getLongestSuffixLength(text1, text2) {
    let length = 0;
    const maxLength = Math.min(text1.length, text2.length);
    const length1 = text1.length - 1;
    const length2 = text2.length - 1;
    while (length < maxLength && text1[length1 - length] === text2[length2 - length]) {
        ++length;
    }
    return length;
}

/**
 * @param {string[]} texts
 * @returns {number}
 */
function getLongestSuffixLengthArray(texts) {
    const count = texts.length;
    switch (count) {
        case 0:
            return 0;
        case 1:
            return texts[0].length;
        default:
        {
            let length = texts[0].length;
            for (let i = 1; i < count; ++i) {
                length = Math.min(length, getLongestSuffixLength(texts[i - 1], texts[i]));
                if (length === 0) { break; }
            }
            return length;
        }
    }
}

/**
 * @param {string[]} texts
 * @returns {number}
 */
function getShortestLength(texts) {
    const count = texts.length;
    if (count === 0) { return 0; }
    let length = texts[0].length;
    for (let i = 1; i < count; ++i) {
        length = Math.min(length, texts[i].length);
    }
    return length;
}

/**
 * @param {string[]} texts
 * @param {number} length
 * @returns {Set<string>}
 */
function getSuffixesOfLength(texts, length) {
    /** @type {Set<string>} */
    const set = new Set();
    for (const text of texts) {
        set.add(text.substring(text.length - length));
    }
    return set;
}

/**
 * @param {string[]} texts
 * @param {number} reductionThreshold
 * @returns {string[]}
 */
function getCommonSuffixes(texts, reductionThreshold) {
    const count = texts.length;
    if (count === 0) { return []; }
    const longestSuffixLength = getLongestSuffixLengthArray(texts);
    const shortestLength = getShortestLength(texts);
    /** @type {?Set<string>} */
    let best = null;
    for (let i = Math.max(1, longestSuffixLength); i <= shortestLength; ++i) {
        const suffixes = getSuffixesOfLength(texts, i);
        if (suffixes.size / count >= reductionThreshold) { break; }
        best = suffixes;
    }
    return best !== null ? [...best] : [];
}
```

Here is a test using the suffixes of `-data`:

```js
const data = ['かったら', 'たら', 'いたら', 'いだら', 'したら', 'ったら', 'ったら', 'ったら', 'んだら', 'んだら', 'んだら', 'じたら', 'したら', '為たら', 'きたら', '来たら', '來たら', 'いったら', 'おうたら', 'こうたら', 'そうたら', 'とうたら', '行ったら', '逝ったら', '往ったら', '請うたら', '乞うたら', '恋うたら', '問うたら', '負うたら', '沿うたら', '添うたら', '副うたら', '厭うたら', 'のたもうたら'];
console.log(getCommonSuffixes(data, 0.5));
// Outputs: [ 'たら', 'だら' ]
```

The `reductionThreshold` is a little hard to name properly, but it basically is a heuristic which is used to tell the function when multiple suffixes stop being useful.
- `1` means "find as many as possible with the length of `getShortestLength(texts)`"
- `0` means "find the longest possible single suffix and stop.
- `N` means "stop creating multiple suffixes once `suffixes.length / texts.length >= reductionThreshold`".

Hard to say what the best value to choose is, but 0.5 should be decent.

**Comment by StefanVukovic99** [comment]
Letting regex handle the heuristic seems to get it down to 40ms - https://github.com/Scrub1492/lesen-tan/pull/11

**Comment by Scrub1492** [comment]
Thanks, this solution seems good to me

File: ext/js/language/language-transformer.js [hunk 1]
```diff
@@ -65,7 +65,9 @@ export class LanguageTransformer {
                     conditionsOut: conditionFlagsOut
                 });
             }
-            transforms2.push({name, rules: rules2});
+            const suffixes = rules.map((rule) => rule.suffixIn);
+            const suffixHeuristic = new RegExp(`(${suffixes.join('|')})$`);
```
Comment (toasted-nutbread) [inline ext/js/language/language-transformer.js:69]: suffixes should be escaped; see escapeRegExp in core/utilities.js

File: ext/js/language/language-transformer.js [hunk 2]
```diff
@@ -133,7 +135,9 @@ export class LanguageTransformer {
         const results = [this._createTransformedText(sourceText, 0, [])];
         for (let i = 0; i < results.length; ++i) {
             const {text, conditions, trace} = results[i];
-            for (const {name, rules} of this._transforms) {
+            for (const {name, rules, suffixHeuristic} of this._transforms) {
```
Comment (toasted-nutbread) [inline ext/js/language/language-transformer.js:138]: Out of curiosity, is there any performance difference if we don't destructure everything up-front? But instead do something like
```js
for (const transform of this._transforms) {
  if (!transform.suffixHeuristic.test(text)) { continue; }
  const {name, rules} = transform;
  ...
}

Comment (Scrub1492) [reply ext/js/language/language-transformer.js]: It seems that this actually improves performances by a small margin

**Review by djahandarie — APPROVED** [review]

**Pull request merged**